text/0101-revamping-the-sdk-performance-api.md

```diff
@@ -0,0 +1,273 @@
+- Start Date: 2023-06-12
+- RFC Type: decision
+- RFC PR: https://github.com/getsentry/rfcs/pull/101
+- RFC Status: draft
+- RFC Driver: [Abhijeet Prasad](https://github.com/abhiprasad)
+
+# Summary
+
+This RFC proposes to revamp the performance API in the SDKs. The new API aims to accomplish the following:
+
+- Align both the internal schemas and top level public API with OpenTelemetry and their SDKs.
+- De-emphasize the concept of transactions from users using a Sentry performance monitoring SDK.
+- Optimize for making sure performance data is always sent to Sentry.
+- Open the SDK up for future work where we support batch span ingestion instead of relying on a transaction solely.
+
+# Planned Work
+
+The RFC proposal comes in two steps.
+
+1. Introduce three new top level methods, `Sentry.startActiveSpan` and `Sentry.startSpan` and their language specific variants as well as `Sentry.setMeasurement`.
+
+This allows us to de-emphasize the concept of hubs, scopes and transactions from users, and instead have them think about just spans. Under the hood, `Sentry.startActiveSpan` and `Sentry.startSpan` should create transactions/spans as appropriate. `Sentry.setMeasurement` is used to abstract away `transaction.setMeasurement` and similar.
```

Comment:
`Sentry.startActiveSpan` should be the default, hence called `Sentry.startSpan`.
Introducing more top-level functions (even undocumented) increases the API surface unnecessarily and adds more confusion.

Reply:
What if we turn it around - instead of having `startActiveSpan` we just have `startSpan` that does by default what `startActiveSpan` does (meaning putting it on the scope) - I suppose this is what people would expect 99% of the time. We could add an argument that prevents setting it on the scope if someone sets it.

Reply:
That's what we do on Mobile, `startTransaction` with a `bindToScope` param, but it's `bindToScope=false` by default.

If this flag is enabled by default, because of single Hub/Scope on Mobile, might be overwriting each other transactions as a side-effect, not sure if it's a good idea, at least on the Mobile world, maybe keeping `bindToScope=false` by default makes more sense in this case.

The other option is that if it's enabled by default, but there's a living transaction in the scope, you just create a span, but maybe this is not the desired behavior and we'd need a way to opt-out, in this case, people would need to force `bindToScope=false`.

Ergonomics isn't great again because they will not know if there's a living transaction in the scope.